tests/test_check_stac_metadata.py

```diff
@@ -159,6 +168,103 @@ def should_save_multiple_directories_validation_results(
+
+    dataset_id = any_dataset_id()
+    version_id = any_dataset_version_id()
+
+    with S3Object(
+        file_object=json_dict_to_file_object(
+            {
+                **deepcopy(MINIMAL_VALID_STAC_OBJECT),
+                "links": [
+                    {"href": f"{base_url}{invalid_child_key}", "rel": "child"},
+                ],
+            }
+        ),
+        bucket_name=staging_bucket_name,
+        key=any_safe_filename(),
+    ) as root_s3_object, S3Object(
+        file_object=json_dict_to_file_object(
+            {
+                **deepcopy(MINIMAL_VALID_STAC_OBJECT),
+                "links": [
+                    {"href": f"{base_url}{invalid_grandchild_key}", "rel": "child"},
+                ],
+            }
+        ),
+        bucket_name=staging_bucket_name,
+        key=invalid_child_key,
+    ) as invalid_child_s3_object, S3Object(
+        file_object=json_dict_to_file_object(deepcopy(MINIMAL_VALID_STAC_OBJECT)),
+        bucket_name=staging_bucket_name,
+        key=invalid_grandchild_key,
+    ) as invalid_grandchild_s3_object:
+        sys.argv = [
+            any_program_name(),
+            f"--metadata-url={root_s3_object.url}",
+            f"--dataset-id={dataset_id}",
+            f"--version-id={version_id}",
+        ]
+
+        with subtests.test(msg="Exit code"):
+            assert main() == 0
+
+        hash_key = f"DATASET#{dataset_id}#VERSION#{version_id}"
+        expected_error = "“{}” links to metadata file in different directory: “{}”"
+        with subtests.test(msg="S3 url validation results"):
+            validation_results_factory_mock.assert_has_calls(
+                [
+                    call(hash_key),
+                    call().save(
+                        root_s3_object.url,
+                        Check.JSON_SCHEMA,
+                        ValidationResult.PASSED,
+                    ),
+                    call().save(
+                        root_s3_object.url,
+                        Check.MULTIPLE_DIRECTORIES,
+                        ValidationResult.FAILED,
+                        details={
+                            "message": expected_error.format(
+                                root_s3_object.url, invalid_child_s3_object.url
+                            )
+                        },
+                    ),
+                    call().save(
+                        invalid_child_s3_object.url,
+                        Check.JSON_SCHEMA,
+                        ValidationResult.PASSED,
+                    ),
+                    call().save(
+                        invalid_child_s3_object.url,
+                        Check.MULTIPLE_DIRECTORIES,
+                        ValidationResult.FAILED,
+                        details={
+                            "message": expected_error.format(
+                                invalid_child_s3_object.url, invalid_grandchild_s3_object.url
```

Comment:
As discussed, the error message should refer to the *root* rather than the *parent* URL. For example, if we have root key `s3://bucket/foo.json`, child `s3://bucket/dir/bar.json` and grandchild `s3://bucket/baz.json`, the grandchild is in the *same* directory as the root and should not provoke an error.